packages/devkit/src/generators/artifact-name-and-directory-utils.spec.ts

```diff
@@ -0,0 +1,863 @@
+          name: 'my-component.component',
+          path: 'apps/app1/src/app/my-component/my-component.component.ts',
+        },
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+      });
+    });
+
+    it('should support receiving a fileName', async () => {
+      addProjectConfiguration(tree, 'app1', {
+        root: 'apps/app1',
+        projectType: 'application',
+      });
+
+      const result = await determineArtifactNameAndDirectoryOptions(tree, {
+        name: 'myComponent',
+        fileName: 'myComponent.component',
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+        artifactName: 'component',
+        callingGenerator: '@my-org/my-plugin:component',
+      });
+
+      expect(result).toStrictEqual({
+        name: 'myComponent',
+        directory: 'apps/app1/src/app/my-component',
+        file: {
+          baseName: 'myComponent.component.ts',
+          name: 'myComponent.component',
+          path: 'apps/app1/src/app/my-component/myComponent.component.ts',
+        },
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+      });
+    });
+
+    it('should support "--pascalCaseFile"', async () => {
+      addProjectConfiguration(tree, 'app1', {
+        root: 'apps/app1',
+        projectType: 'application',
+      });
+
+      const result = await determineArtifactNameAndDirectoryOptions(tree, {
+        name: 'myComponent',
+        pascalCaseFile: true,
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+        artifactName: 'component',
+        callingGenerator: '@my-org/my-plugin:component',
+      });
+
+      expect(result).toStrictEqual({
+        name: 'myComponent',
+        directory: 'apps/app1/src/app/my-component',
+        file: {
+          baseName: 'MyComponent.ts',
+          name: 'MyComponent',
+          path: 'apps/app1/src/app/my-component/MyComponent.ts',
+        },
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+      });
+    });
+
+    it('should support "--pascalCaseDirectory"', async () => {
+      addProjectConfiguration(tree, 'app1', {
+        root: 'apps/app1',
+        projectType: 'application',
+      });
+
+      const result = await determineArtifactNameAndDirectoryOptions(tree, {
+        name: 'myComponent',
+        pascalCaseDirectory: true,
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+        artifactName: 'component',
+        callingGenerator: '@my-org/my-plugin:component',
+      });
+
+      expect(result).toStrictEqual({
+        name: 'myComponent',
+        directory: 'apps/app1/src/app/MyComponent',
+        file: {
+          baseName: 'my-component.ts',
+          name: 'my-component',
+          path: 'apps/app1/src/app/MyComponent/my-component.ts',
+        },
+        project: 'app1',
+        nameAndDirectoryFormat: 'derived',
+      });
+    });
+
+    it('should support receiving a different file extension', async () => {
```

Comment:
Do we need to support file extensions? The component generators for both Angular and React only need to know what the `fileName` is (e.g. `my-component` vs `MyComponent`). The extensions are in the templates passed to `generateFiles`.

I don't think we need to expand the `file` entry, it can just be a string like:

```json
{
  name: 'myComponent',
  directory: apps/app1/src/app/MyComponent',
  file: 'my-component'
  project: 'app1',
  nameAndDirectoryFormat: 'derived',
}
```

The rest can be computed from those two values.

---

One small thing, I think `artifactName` should be `artifactType`, same way that existing `determineProjectNameAndRootOptions` function takes in `projectType`. And then the returned `name` should be `artifactName`, so the two functions are consistent.

```json
{
  artifactName: 'myComponent',
  directory: apps/app1/src/app/MyComponent',
  file: 'my-component'
  project: 'app1',
  nameAndDirectoryFormat: 'derived',
}
```

Reply:
Good points. I agree and by doing that we could even leave the handling of the suffix out of the helper. That's something the consumer can compose with the information returned.

Reply:
As discussed offline, the helper needs the file extension. We'll simplify some bits of the return type and rename what was suggested.